plugin/execCommand.go

```diff
@@ -0,0 +1,128 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+)
+
+// ExecCommandImpl implements the plugin.ExecCommand interface.
+// Use plugin.NewExecCommand to create instances of these objects.
+type ExecCommandImpl struct {
+	ctx           context.Context
+	outputCh      chan ExecOutputChunk
+	stdout        *OutputStream
+	stderr        *OutputStream
+	exitCodeCh    chan int
+	exitCodeErrCh chan error
+}
+
+// NewExecCommand creates a new ExecCommandImpl object whose
+// lifetime is tied to the passed-in execution context.
+func NewExecCommand(ctx context.Context) *ExecCommandImpl {
+	cmd := &ExecCommandImpl{
+		ctx: ctx,
+		exitCodeCh: make(chan int, 1),
+		exitCodeErrCh: make(chan error, 1),
+	}
+
+	// Create the output streams
+	cmd.outputCh = make(chan ExecOutputChunk)
+	closer := &multiCloser{ch: cmd.outputCh, countdown: 2}
+	cmd.stdout = &OutputStream{ctx: cmd.ctx, id: Stdout, ch: cmd.outputCh, closer: closer}
+	cmd.stderr = &OutputStream{ctx: cmd.ctx, id: Stderr, ch: cmd.outputCh, closer: closer}
+
+	// Ensure that the output streams are closed when the context
+	// is cancelled. This guarantees that callers won't be blocked
+	// when they are streaming our command's output.
+	go func() {
```

Comment:
Note that this cleanup cleanly handles edge-cases like a failing Stop call leaving dangling `Exec`-related goroutines (b/c the command would still be running, so the output streams would still be open and hence the caller blocked).

It might be worthwhile (in the future) to add some retry logic to `SetStopFunc` in case a Stop fails due to some connection error, though I don't know how likely that scenario is (or if it's even possible).